File: ipysimulate/control.py
import traitlets
import ipywidgets
import threading
import time
import ipysimulate
from .tools import make_list
from .parameters import Range, IntRange, Values

# See js/lib/control.js for the frontend counterpart to this file.
semver_range = "~" + ipysimulate.__version__  # Retrieve version

# Prepare parameter classes
range_types = (Range, )
intrange_types = (IntRange, )
value_types = (Values, )
try:
    import agentpy as ap
    if ap.__version__ >= '0.0.8':
        range_types += (ap.Range, )
        intrange_types += (ap.IntRange, )
        value_types += (ap.Values, )
except ImportError as e:
    pass


@ipywidgets.register
class Control(ipywidgets.DOMWidget):
    """ Control panel widget for an interactive simulation.

    Arguments:
        model:
            A :ref:`simulation model <simulation_model>` with discrete steps.
        parameters (dict, optional):
            Dictionary of parameter names and values (default None).
            Entries of type :class:`Range`, :class:`IntRange`,
            and :class:`Values` will be displayed as interactive widgets.
        variables (str of list of str, optional):
            Model attributes to display in the control panel (default None).
    """

    # Traitlet declarations ------------------------------------------------- #
    
    _view_name = traitlets.Unicode('ControlView').tag(sync=True)
    _view_module = traitlets.Unicode('ipysimulate').tag(sync=True)
    _view_module_version = traitlets.Unicode(semver_range).tag(sync=True)
    _model_name = traitlets.Unicode('ControlModel').tag(sync=True)
    _model_module = traitlets.Unicode('ipysimulate').tag(sync=True)
    _model_module_version = traitlets.Unicode(semver_range).tag(sync=True)

    is_running = traitlets.Bool(False).tag(sync=True)
    do_reset = traitlets.Bool(False).tag(sync=True)

    _variables = traitlets.Dict().tag(sync=True)
    parameters = traitlets.Dict().tag(sync=True)
    data_paths = traitlets.List().tag(sync=True)
    _pwidgets = traitlets.List().tag(sync=True)
    t = traitlets.Integer(0).tag(sync=True)

    name = traitlets.Unicode().tag(sync=True)

    # Initiation - Don't start any threads here ----------------------------- #
    
    def __init__(self, model, parameters=None, variables=None):
        super().__init__()  # Initiate front-end
        self.on_msg(self._handle_button_msg)  # Handle front-end messages
        self.thread = None  # Placeholder for simulation threads
        self._pre_pwidgets = []
        self._pdtypes = {}

        self.parameters = {}
        if parameters:
            for k, v in parameters.items():
                if isinstance(v, value_types):
                    self._create_select(k, v)
                    self.parameters[k] = v.vdef
                elif isinstance(v, intrange_types):
                    self._create_slider(k, v, int_slider=True)
                    self.parameters[k] = v.vdef
                elif isinstance(v, range_types):
                    self._create_slider(k, v)
                    self.parameters[k] = v.vdef
                else:
                    self.parameters[k] = v

        self._pwidgets = self._pre_pwidgets
        self.model = model
        self.model.set_parameters(self.parameters)

        self._callbacks = []
        self._var_keys = make_list(variables)
        self._variables = {k: None for k in self._var_keys}

        self.charts = []

    # Callbacks ------------------------------------------------------------- #

    def add_callback(self, func, *args, **kwargs):
        self._callbacks.append((func, args, kwargs))

    # Parameter widgets ----------------------------------------------------- #

    def _create_slider(self, k, v, int_slider=False):
        pwidget = {
            'name': k,
            'type': 'slider',
            'vmin': v.vmin,
            'vmax': v.vmax,
            'vdef': v.vdef
        }
        if int_slider:
            pwidget['step'] = max([1, int((v.vmax - v.vmin) / 100)])
            self._pdtypes[k] = int
        else:
            pwidget['step'] = (v.vmax - v.vmin) / 100
            self._pdtypes[k] = float
        self._pre_pwidgets.append(pwidget)

    def _create_select(self, k, v):
        pwidget = {
            'name': k,
            'type': 'select',
            'values': v.values,
            'vdef': v.vdef
        }
        # TODO Better way to infer dtypes
        self._pdtypes[k] = type(v.values[0])
        self._pre_pwidgets.append(pwidget)

    # Methods to be called from the front-end ------------------------------- #

    def _handle_button_msg(self, _, content, buffers):
        """ Handles messages from the front-end 
        by calling method of same name as msg. """
        getattr(self, content.get('event', ''))(**content)

    def update_parameter(self, k, v):
        self.model.p[k] = self._pdtypes[k](v)

    def setup_simulation(self, **kwargs):
        """ Call model setup. """
        self.thread = threading.Thread(target=self.run_setup)
        self.thread.start()
        
    def continue_simulation(self, **kwargs):
        """ Start background thread that runs simulation. """
        self.thread = threading.Thread(target=self.run_simulation)
        self.thread.start()
        
    def increment_simulation(self, **kwargs):
        """ Do a single simulation step. """
        self.thread = threading.Thread(target=self.run_step)
        self.thread.start()
        
    def reset_simulation(self, **kwargs):
        """ Reset graphs and simulation. """
        self.thread = threading.Thread(target=self.reset)
        self.thread.start()
        
    # Methods to be called only within threads ------------------------------ #

    def sync_data(self):
        """ Retrieve new data from simulation and send it to front-end. """
        self._variables = {k: getattr(self.model, k) for k in self._var_keys}
        for chart in self.charts:
            chart.sync_data()
        for callback, args, kwargs in self._callbacks:
            callback(*args, **kwargs)

    def reset(self):
        """ Reset simulation by clearing front-end data,
        calling `model.sim_reset()`, and sending initial data to front-end."""
        for chart in self.charts:
            chart.reset_data()
        self.run_setup()  # Reset backend model by calling setup again

    def run_setup(self):
        """ Initiate simulation by calling `model.sim_setup()`
        and sending initial data to front-end. """
        self.model.sim_setup()
        self.t = self.model.t
        self.sync_data()
    
    def run_step(self):
        """ Run a single simulation step by calling `model.sim_step()`,
        and sending new data to front-end. """
        self.model.sim_step()
        self.t = self.model.t
        self.sync_data()
        
    def run_simulation(self):
        """ Start or continue the simulation by repeatedly calling
        :func:`Control.run_single_step` as long as `model.active` is True. """
        self.is_running = True
        if 'fps' in self.model.p:
            while self.model.running:
                start = time.time()
                self.run_step()
                wait = 1 / self.model.p.fps + start - time.time()
                if wait > 0:
                    time.sleep(wait)
                if not self.is_running:
                    break
        else:
            while self.model.running:
                self.run_step()
                if not self.is_running:
                    break
        self.is_running = False
        if self.do_reset:
            self.reset_simulation()
            self.do_reset = False

File: js/lib/control.js
// Interactive control panel for simulations
// See control.py for the kernel counterpart to this file.

var widgets = require('@jupyter-widgets/base');
var semver_range = require('../package.json').version;
require('./control.css');
require('lodash');


var ControlModel = widgets.DOMWidgetModel.extend({
    defaults: _.extend(widgets.DOMWidgetModel.prototype.defaults(), {
        _model_name : 'ControlModel',
        _view_name : 'ControlView',
        _model_module : 'ipysimulate',
        _view_module : 'ipysimulate',
        _model_module_version : semver_range,
        _view_module_version : semver_range,
    }),

    initialize: function (attributes, options) {
        widgets.DOMWidgetModel.prototype
            .initialize.call(this, attributes, options);
        this.charts = [];
    },

});


var ControlView = widgets.DOMWidgetView.extend({
        
    // React to control button events -------------------------------------- //
    click_play: function(inputEvent) {
        // Continue simulation if model is idle, else pause.
        if (this.model.get('is_running') === false) {
            this.send({event: 'continue_simulation'});
        } else {  
            this.model.set({'is_running':false});
            this.touch(); // Sync JS & Python
        }
    },
    
    click_step: function(inputEvent) {
        // Run single step if model is idle.
        if (this.model.get('is_running') == false) {
            this.send({event: 'increment_simulation'});
        }
    },
    
    click_redo: function(inputEvent) {
        // Set to reset after end of step, if model is runnning.
        // Reset through function call if model is idle.
        if (this.model.get('is_running') == true) {
            this.model.set({'do_reset':true, 'is_running':false});
            this.touch(); // Sync JS & Python 
        } else {
            this.send({event: 'reset_simulation'});
        }
    },

    click_info: function(inputEvent) {
        // TODO Do something useful with this button
        // alert(this.model.get(''))
    },

    param_label_change: function(name, value, label) {
        // For updates after release
        label.textContent = value
        this.param_change(name, value)
    },

    param_change: function(name, value) {
        // For updates after release
        this.send({
            event: 'update_parameter',
            k: name,
            v: value,
        })
    },

    // Render control interface -------------------------------------------- //
    render: function() {

        // Handle traitlet changes
        this.model.on('change:is_running', this.is_running_changed, this);
        this.model.on('change:_variables', this.variables_changed, this);
        
        // Control interface ----------------------------------------------- //
        this.control = document.createElement("div");
        this.control.className = 'ipysimulate-control'
        this.el.appendChild(this.control);

        // Button row
        this.control_line = document.createElement("div");
        this.control_line.className = "button-row"
        this.control.appendChild(this.control_line);
        
        // Play/Pause button
        this.play_button = document.createElement("BUTTON");
        this.play_button.className = "ctr_btn fa fa-play";
        this.play_button.setAttribute('title', 'Run simulation');
        this.control_line.appendChild(this.play_button);
        this.play_button.addEventListener("click",
            (inputEvent => this.click_play()), false);
        
        // Step button
        this.step_button = document.createElement("BUTTON");
        this.step_button.className = "ctr_btn fa fa-step-forward";
        this.step_button.setAttribute('title', 'Run single step')
        this.control_line.appendChild(this.step_button);
        this.step_button.addEventListener("click",
            (inputEvent => this.click_step()), false);

        // Restart button
        this.redo_button = document.createElement("BUTTON");
        this.redo_button.className = "ctr_btn last fa fa-repeat";
        this.redo_button.setAttribute('title', 'Reset simulation');
        this.control_line.appendChild(this.redo_button);
        this.redo_button.addEventListener("click",
            (inputEvent => this.click_redo()), false);

        // Variable widgets ------------------------------------------------ //
        this.var_displays = {}
        let variables = this.model.get('_variables');
        for (const [key, value] of Object.entries(variables)) {

            // Create new row
            let row = document.createElement("div");
            this.control.appendChild(row);
            row.className = "row";

            // Create labels
            let label = document.createElement("span");
            label.textContent = key + ' ';
            label.className = "plabel";
            row.appendChild(label);

            let label2 = document.createElement("span");
            label2.textContent = value
            row.appendChild(label2);
            this.var_displays[key] = label2
        }

        // Parameter widgets ----------------------------------------------- //

        var i;
        let pwidgets = this.model.get('_pwidgets')
        for (i = 0; i < pwidgets.length; i++) {
            let pwidget = pwidgets[i]
            let row = document.createElement("div");
            row.className = "row"

            let label = document.createElement("span");
            label.className = "plabel"
            label.textContent = pwidget.name + " "

            row.appendChild(label)
            this.control.appendChild(row);

            switch(pwidget.type)
            {
            // Slider widget ----------------------------------------------- //
                case 'slider':

                    let label3 = document.createElement("span");
                    label3.textContent = pwidget.vdef.toString()
                    row.appendChild(label3)

                    let slider = document.createElement("input");
                    slider.className = "slider"
                    slider.setAttribute('id', 'myRange')
                    slider.setAttribute('type', 'range')
                    slider.setAttribute('step', pwidget.step)
                    slider.setAttribute('min', pwidget.vmin.toString())
                    slider.setAttribute('max', pwidget.vmax.toString())
                    slider.setAttribute('value', pwidget.vdef.toString())
                    slider.addEventListener("change",
                            inputEvent => this.param_label_change(
                                pwidget.name, slider.value, label3));
                    slider.addEventListener("input",
                            inputEvent => this.param_label_change(
                                pwidget.name, slider.value, label3));
                    row.appendChild(slider)

                break;

            // Select widget ----------------------------------------------- //
            case 'select':
                let options = pwidget.values;
                let select = document.createElement("select");
                select.setAttribute('class', 'select')

                //Create and append the options
                var j;
                for (j = 0; j < options.length; j++) {
                    let option = document.createElement("option");
                    option.value = options[j];
                    option.text = options[j];
                    select.appendChild(option);
                }
                select.addEventListener("change",
                        inputEvent => this.param_change(pwidget.name, select.value));

                row.appendChild(select);
                break;
            }
        }

        // Setup ----------------------------------------------------------- //
        this.send({event: 'setup_simulation'});
    },

    // React to tratilet changes ------------------------------------------- //
    is_running_changed: function() {
        if (this.model.get('is_running') === true) {
            this.play_button.className = "ctr_btn fa fa-pause";
            this.play_button.setAttribute('title', 'Pause simulation');
        } else {
            this.play_button.className = "ctr_btn fa fa-play";
            this.play_button.setAttribute('title', 'Run simulation');
        }
    },

    variables_changed: function() {
        let variables = this.model.get('_variables');
        for (const [key, value] of Object.entries(variables)) {
			this.var_displays[key].textContent = value
		}
        //this.output2.textContent = this.model.get('t');
    },
    
    reset_changed: function() {
        this.model.data = []
    }

});


module.exports = {
    ControlModel: ControlModel,
    ControlView: ControlView,
};


